File: backend/src/main/java/com/kiwi/common/types/Password.java
package com.kiwi.common.types;

import java.util.regex.Pattern;

public record Password(String value) {
    private static final Pattern PASSWORD_REGEX =
            Pattern.compile("^(?=.*[a-z])(?=.*[A-Z])(?=.*\\d)[A-Za-z\\d@$!%*?&]{8,}$");

    public Password {
        if (!isValid(value)) throw new IllegalArgumentException("Invalid password format");
    }

    private boolean isValid(String value) {
        return PASSWORD_REGEX.matcher(value).matches();
    }
}

File: mobile/app/src/main/java/com/bellako/kiwi/login/Password.kt
package com.bellako.kiwi.login

@JvmInline
value class Password private constructor(val value: String) {
    companion object {
        private val PASSWORD_REGEX = Regex(
            "^(?=.*[a-z])(?=.*[A-Z])(?=.*\\d)(?=.*[@$!%*?&])[A-Za-z\\d@$!%*?&]{8,}$"
        )

        fun isValid(password: String): Boolean {
            return PASSWORD_REGEX.matches(password)
        }

        fun of(password: String): Result<Password> {
            return if (isValid(password)) {
                Result.success(Password(password))
            } else {
                Result.failure(IllegalArgumentException("Invalid password format"))
            }
        }
    }
}